### DISCUSSION TITLE
Calling `notFound()` with (top level) `<Suspense>` boundary is causing status code 200 instead of 404

### DISCUSSION THREAD

**Discussion by aronwoost** [vercel/next.js#70488] [Help]
### Link to the code that reproduces this issue

https://github.com/aronwoost/nextjs-suspense-notfound-issue

### To Reproduce

1. Checkout the repo
2. Run `npm install`
3. Run `npm run dev`
4. Open http://localhost:3000/product/10
5. "not found" page opens with status code 200

### Current vs. Expected behavior

Current: "not found" page opens with status code 200
Expected: "not found" page opens with status code 404

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.5.0: Wed May  1 20:19:05 PDT 2024; root:xnu-10063.121.3~5/RELEASE_ARM64_T8112
  Available memory (MB): 16384
  Available CPU cores: 8
Binaries:
  Node: 18.19.1
  npm: 10.2.4
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 14.2.13 // Latest available version is detected (14.2.13).
  eslint-config-next: 14.2.13
  react: 18.3.1
  react-dom: 18.3.1
  typescript: 5.6.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Not sure

### Which stage(s) are affected? (Select all that apply)

next dev (local), next build (local), next start (local), Vercel (Deployed), Other (Deployed)

### Additional context

Find more details in the `README` here: https://github.com/aronwoost/nextjs-suspense-notfound-issue

**Comment by mmdrn** [comment]
I have same issue:
![image](https://github.com/user-attachments/assets/4ce95b42-c2bd-4a51-bf32-3a4a0cc868da)

When I open the /product/[id], the not found page will be triggered.

not found code:
![image](https://github.com/user-attachments/assets/217e0610-a2e7-4664-9299-43afe1d293c0)

**Comment by abhi12299** [comment]
I believe this is not a bug - when you wrap a component tree in `<Suspense>`, only the fallback is rendered on the server and the `children` are streamed in later. (PPR) This means that during the initial render of the page (when fallback is rendered), nothing went wrong on the server, and it responded with a `200` status code.

**Comment by aronwoost** [comment]
Thanks for jumping in @abhi12299

1. My previous understanding was, that calling `notFound()` is similar to an error, but not actually an error. But reading [the docs](https://nextjs.org/docs/app/api-reference/functions/not-found) again I understand better now, that it's a "real" error. And might indeed be catched by a `<Suspense>`.
2. Why does `http://localhost:3000/not-existing-url` behaves different and does respond with a 404?
3. And of course: How to work around it and still get a 404 with `notFound()` and `<Suspense>`? 😊

**Comment by abhi12299** [comment]
1. Yes, `notFound` just throws an error with some metadata attached to it so that it can be identified by the error boundary.
2. `http://localhost:3000/not-existing-url` behaves differently because the server is able to recognise that this URL does not exist (via file system routing) and is able to set the status code even before the render.
3. Use `notFound()` outside of `<Suspense>` to make sure that the `NOT_FOUND_ERROR` is captured during the initial render of the page, so that the server can respond with the relevant status code.

**Reply by gormonn** [reply]
3. This seems like a workaround, not a solution. Especially since it will slow down all the fun of using Suspense.

But... It looks like the Streaming concept itself doesn't allow us to do this as we expect. Because the first response will always be 200, until the end of the stream.
